pkg/volume/glusterfs/glusterfs.go

```diff
@@ -141,12 +141,51 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
 	if kubeClient == nil {
 		return nil, fmt.Errorf("glusterfs: failed to get kube client to initialize mounter")
 	}
+
 	ep, err := kubeClient.Core().Endpoints(ns).Get(epName, metav1.GetOptions{})
-	if err != nil {
-		glog.Errorf("glusterfs: failed to get endpoints %s[%v]", epName, err)
-		return nil, err
+	if err != nil && errors.IsNotFound(err) {
+		glog.Errorf("glusterfs: failed to get endpoint %s[%v]", epName, err)
+		if spec != nil && spec.PersistentVolume.Annotations["kubernetes.io/createdby"] == "heketi-dynamic-provisioner" {
+			class, err := volutil.GetClassForVolume(plugin.host.GetKubeClient(), spec.PersistentVolume)
+			if err != nil {
+				return nil, fmt.Errorf("glusterfs: failed to get storageclass when recreating endpoint/service: %v", err)
+			}
+
+			cfg, err := parseClassParameters(class.Parameters, plugin.host.GetKubeClient())
+			if err != nil {
+				return nil, fmt.Errorf("glusterfs: failed to parse storageclass parameter %v when recreating endpoint/service", err)
+			}
+
+			scConfig := *cfg
+			cli := gcli.NewClient(scConfig.url, scConfig.user, scConfig.secretValue)
+			if cli == nil {
+				return nil, fmt.Errorf("glusterfs: failed to create glusterfs rest client when recreating endpoint/service ")
+			}
+
+			volumeID := dstrings.TrimPrefix(source.Path, volPrefix)
+			volInfo, err := cli.VolumeInfo(volumeID)
+			if err != nil {
+				return nil, fmt.Errorf("glusterfs: failed to fetch volumeID: %v when recreating endpoint/service", err)
+			}
+
+			endpointIPs, err := getClusterNodes(cli, volInfo.Cluster)
+			if err != nil {
+				return nil, fmt.Errorf("glusterfs: failed to get clusterIPs: %v when recreating endpoint/service", err)
+			}
+
+			// Give an attempt to recreate endpoint/service.
+			claim := spec.PersistentVolume.Spec.ClaimRef.Name
+			_, _, err = plugin.createEndpointService(ns, epName, endpointIPs, claim)
+			if err != nil {
+				glog.Errorf("glusterfs: failed to recreate endpoint/service: %v", err)
```

Comment:
There could be a race condition when two PVs are used simultaneously and yet one PV creates the endpoint before the other.